### ISSUE TITLE
How can I pass two identical pciDevices?

### ISSUE THREAD

**Issue by alizshow** [kubevirt/kubevirt#7755]
I have two identical GPUs with the same vendor/product id and in the same IOMMU group and I wish to pass them both to a single VM.
The Kubevirt-gpu-device-plugin provided by Nvidia detects one of them only, so I tried adding and naming these under different names in the `permittedHostDevices` of the config (although there is no way for me to differentiate between the two) and passing them to the VM directly.

```
permittedHostDevices:
      pciHostDevices:
      - pciVendorSelector: 10de:1e07
        resourceName: example.com/2080_Ti
      - pciVendorSelector: 10de:1e07
        resourceName: example.com/2080_Ti2
```
```
hostDevices:
    - deviceName: example.com/2080_Ti
      name: gpu1
    - deviceName: example.com/2080_Ti2
      name: gpu2
```

Which is somewhat pointless since only one of the devices show up on the node description with non-empty capacity:
```
Capacity:
  example.com/2080_Ti:                         0
  example.com/2080_Ti2:                        1
```

I am using KubeVirt v0.41.3

TL;DR
Have two GPUs in the same IOMMU group and want to pass them to a single VM, which should be possible.

**Comment by alicefr** [comment]
hi @alizshow there has been a couple of issues in updating the device plugin. Can you please provide the virt-handler logs?
One thing you can try is to restart the virt-handler pod by deleting the pod and check if this was updated. Another source of confusion for the device plugin is if you using the KubeVirt configmap together with the KubeVirt CRD. Is it your case?

**Comment by alizshow** [comment]
virt-handler logs regarding the GPU device plugin:
```
{"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A]","pos":"device_controller.go:225","timestamp":"2022-05-16T13:21:52.747912Z"}
{"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A","pos":"device_controller.go:92","timestamp":"2022-05-16T13:21:52.747965Z"}
{"component":"virt-handler","level":"info","msg":"nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-16T13:21:52.748975Z"}
{"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: example.com/2080_Ti","pos":"device_controller.go:92","timestamp":"2022-05-17T06:03:30.353681Z"}
{"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:225","timestamp":"2022-05-17T06:03:30.353714Z"}
{"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A]","pos":"device_controller.go:226","timestamp":"2022-05-17T06:03:30.353791Z"}
{"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T06:03:30.356123Z"}
{"component":"virt-handler","level":"info","msg":"device socket file for device example.com/2080_Ti was removed, kubelet probably restarted.","pos":"pci_device.go:308","timestamp":"2022-05-17T06:03:30.366147Z"}
{"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T06:03:31.384759Z"}
{"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: example.com/2080_Ti2","pos":"device_controller.go:92","timestamp":"2022-05-17T06:07:54.507215Z"}
{"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [example.com/2080_Ti2]","pos":"device_controller.go:225","timestamp":"2022-05-17T06:07:54.507244Z"}
{"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:226","timestamp":"2022-05-17T06:07:54.507280Z"}
{"component":"virt-handler","level":"info","msg":"example.com/2080_Ti2 device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T06:07:54.508613Z"}
{"component":"virt-handler","level":"info","msg":"device socket file for device example.com/2080_Ti2 was removed, kubelet probably restarted.","pos":"pci_device.go:308","timestamp":"2022-05-17T06:07:54.514153Z"}
{"component":"virt-handler","level":"info","msg":"example.com/2080_Ti2 device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T06:07:55.537031Z"}
{"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:225","timestamp":"2022-05-17T09:04:24.645954Z"}
{"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: example.com/2080_Ti","pos":"device_controller.go:92","timestamp":"2022-05-17T09:04:24.645976Z"}
{"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T09:04:24.648323Z"}
{"component":"virt-handler","kind":"","level":"info","msg":"VMI is in phase: Scheduled\n","name":"ozma-9x614ntakq","namespace":"vm","pos":"vm.go:1537","timestamp":"2022-05-17T09:05:48.208003Z","uid":"6f3b34bb-92e3-448c-9de2-959482d94649"}
{"component":"virt-handler","kind":"Domain","level":"info","msg":"Domain status: Running, reason: Unknown\n","name":"ozma-9x614ntakq","namespace":"vm","pos":"vm.go:1542","timestamp":"2022-05-17T09:05:48.208017Z","uid":"6f3b34bb-92e3-448c-9de2-959482d94649"}
{"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:226","timestamp":"2022-05-17T09:11:29.950320Z"}
{"component":"virt-handler","level":"info","msg":"refreshed device plugins for permitted/forbidden host devices","pos":"device_controller.go:224","timestamp":"2022-05-17T09:11:59.208049Z"}
{"component":"virt-handler","level":"info","msg":"enabled device-pluings for: []","pos":"device_controller.go:225","timestamp":"2022-05-17T09:11:59.208064Z"}
{"component":"virt-handler","level":"info","msg":"disabled device-pluings for: []","pos":"device_controller.go:226","timestamp":"2022-05-17T09:11:59.208071Z"}
{"component":"virt-handler","level":"info","msg":"Updating cluster config from KubeVirt to resource version '170720807'","pos":"config-map.go:516","timestamp":"2022-05-17T09:15:57.266360Z"}
{"component":"virt-handler","level":"info","msg":"Updating cluster config from KubeVirt to resource version '170779681'","pos":"config-map.go:516","timestamp":"2022-05-17T11:03:48.820802Z"}
{"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:225","timestamp":"2022-05-17T11:06:02.576941Z"}
{"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [example.com/2080_Ti2]","pos":"device_controller.go:226","timestamp":"2022-05-17T11:06:02.576961Z"}
{"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: example.com/2080_Ti","pos":"device_controller.go:92","timestamp":"2022-05-17T11:06:02.576962Z"}
{"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:06:02.578046Z"}
{"component":"virt-handler","level":"info","msg":"device socket file for device example.com/2080_Ti was removed, kubelet probably restarted.","pos":"pci_device.go:308","timestamp":"2022-05-17T11:06:02.594368Z"}
{"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:06:03.601403Z"}
{"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A]","pos":"device_controller.go:225","timestamp":"2022-05-17T11:12:49.072044Z"}
{"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A","pos":"device_controller.go:92","timestamp":"2022-05-17T11:12:49.072034Z"}
{"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:226","timestamp":"2022-05-17T11:12:49.072082Z"}
{"component":"virt-handler","level":"info","msg":"nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:12:49.074346Z"}
{"component":"virt-handler","level":"info","msg":"device socket file for device nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A was removed, kubelet probably restarted.","pos":"pci_device.go:308","timestamp":"2022-05-17T11:12:49.086378Z"}
{"component":"virt-handler","level":"info","msg":"nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:12:50.108700Z"}
{"component":"virt-handler","level":"info","msg":"set verbosity to 2","pos":"virt-handler.go:427","timestamp":"2022-05-17T11:12:57.452080Z"}
{"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:225","timestamp":"2022-05-17T11:37:17.990751Z"}
{"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: example.com/2080_Ti","pos":"device_controller.go:92","timestamp":"2022-05-17T11:37:17.990724Z"}
{"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A]","pos":"device_controller.go:226","timestamp":"2022-05-17T11:37:17.990793Z"}
{"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:37:17.993336Z"}
{"component":"virt-handler","level":"info","msg":"device socket file for device example.com/2080_Ti was removed, kubelet probably restarted.","pos":"pci_device.go:308","timestamp":"2022-05-17T11:37:17.998289Z"}
{"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:37:19.001290Z"}
```

I'm using the KubeVirt CRD, the configmaps present are
```
kube-root-ca.crt
kubevirt-ca
kubevirt-install-strategy-jm7m2
```
which I do not think overlap with the CRD.

so @alicefr, you don't think the problem has any relevance to the fact that my GPUs are in the same IOMMU group and are identical? I have many other nodes in the cluster (that do not have this condition) and they (and their GPUs) are behaving very nicely.

**Comment by alicefr** [comment]
> virt-handler logs regarding the GPU device plugin:
> 
> ```
> {"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A]","pos":"device_controller.go:225","timestamp":"2022-05-16T13:21:52.747912Z"}
> {"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A","pos":"device_controller.go:92","timestamp":"2022-05-16T13:21:52.747965Z"}
> {"component":"virt-handler","level":"info","msg":"nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-16T13:21:52.748975Z"}
> {"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: example.com/2080_Ti","pos":"device_controller.go:92","timestamp":"2022-05-17T06:03:30.353681Z"}
> {"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:225","timestamp":"2022-05-17T06:03:30.353714Z"}
> {"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A]","pos":"device_controller.go:226","timestamp":"2022-05-17T06:03:30.353791Z"}
> {"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T06:03:30.356123Z"}
> {"component":"virt-handler","level":"info","msg":"device socket file for device example.com/2080_Ti was removed, kubelet probably restarted.","pos":"pci_device.go:308","timestamp":"2022-05-17T06:03:30.366147Z"}
> {"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T06:03:31.384759Z"}
> {"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: example.com/2080_Ti2","pos":"device_controller.go:92","timestamp":"2022-05-17T06:07:54.507215Z"}
> {"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [example.com/2080_Ti2]","pos":"device_controller.go:225","timestamp":"2022-05-17T06:07:54.507244Z"}
> {"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:226","timestamp":"2022-05-17T06:07:54.507280Z"}
> {"component":"virt-handler","level":"info","msg":"example.com/2080_Ti2 device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T06:07:54.508613Z"}
> {"component":"virt-handler","level":"info","msg":"device socket file for device example.com/2080_Ti2 was removed, kubelet probably restarted.","pos":"pci_device.go:308","timestamp":"2022-05-17T06:07:54.514153Z"}
> {"component":"virt-handler","level":"info","msg":"example.com/2080_Ti2 device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T06:07:55.537031Z"}
> {"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:225","timestamp":"2022-05-17T09:04:24.645954Z"}
> {"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: example.com/2080_Ti","pos":"device_controller.go:92","timestamp":"2022-05-17T09:04:24.645976Z"}
> {"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T09:04:24.648323Z"}
> {"component":"virt-handler","kind":"","level":"info","msg":"VMI is in phase: Scheduled\n","name":"ozma-9x614ntakq","namespace":"vm","pos":"vm.go:1537","timestamp":"2022-05-17T09:05:48.208003Z","uid":"6f3b34bb-92e3-448c-9de2-959482d94649"}
> {"component":"virt-handler","kind":"Domain","level":"info","msg":"Domain status: Running, reason: Unknown\n","name":"ozma-9x614ntakq","namespace":"vm","pos":"vm.go:1542","timestamp":"2022-05-17T09:05:48.208017Z","uid":"6f3b34bb-92e3-448c-9de2-959482d94649"}
> {"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:226","timestamp":"2022-05-17T09:11:29.950320Z"}
> {"component":"virt-handler","level":"info","msg":"refreshed device plugins for permitted/forbidden host devices","pos":"device_controller.go:224","timestamp":"2022-05-17T09:11:59.208049Z"}
> {"component":"virt-handler","level":"info","msg":"enabled device-pluings for: []","pos":"device_controller.go:225","timestamp":"2022-05-17T09:11:59.208064Z"}
> {"component":"virt-handler","level":"info","msg":"disabled device-pluings for: []","pos":"device_controller.go:226","timestamp":"2022-05-17T09:11:59.208071Z"}
> {"component":"virt-handler","level":"info","msg":"Updating cluster config from KubeVirt to resource version '170720807'","pos":"config-map.go:516","timestamp":"2022-05-17T09:15:57.266360Z"}
> {"component":"virt-handler","level":"info","msg":"Updating cluster config from KubeVirt to resource version '170779681'","pos":"config-map.go:516","timestamp":"2022-05-17T11:03:48.820802Z"}
> {"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:225","timestamp":"2022-05-17T11:06:02.576941Z"}
> {"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [example.com/2080_Ti2]","pos":"device_controller.go:226","timestamp":"2022-05-17T11:06:02.576961Z"}
> {"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: example.com/2080_Ti","pos":"device_controller.go:92","timestamp":"2022-05-17T11:06:02.576962Z"}
> {"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:06:02.578046Z"}
> {"component":"virt-handler","level":"info","msg":"device socket file for device example.com/2080_Ti was removed, kubelet probably restarted.","pos":"pci_device.go:308","timestamp":"2022-05-17T11:06:02.594368Z"}
> {"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:06:03.601403Z"}
> {"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A]","pos":"device_controller.go:225","timestamp":"2022-05-17T11:12:49.072044Z"}
> {"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A","pos":"device_controller.go:92","timestamp":"2022-05-17T11:12:49.072034Z"}
> {"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:226","timestamp":"2022-05-17T11:12:49.072082Z"}
> {"component":"virt-handler","level":"info","msg":"nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:12:49.074346Z"}
> {"component":"virt-handler","level":"info","msg":"device socket file for device nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A was removed, kubelet probably restarted.","pos":"pci_device.go:308","timestamp":"2022-05-17T11:12:49.086378Z"}
> {"component":"virt-handler","level":"info","msg":"nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:12:50.108700Z"}
> {"component":"virt-handler","level":"info","msg":"set verbosity to 2","pos":"virt-handler.go:427","timestamp":"2022-05-17T11:12:57.452080Z"}
> {"component":"virt-handler","level":"info","msg":"enabled device-pluings for: [example.com/2080_Ti]","pos":"device_controller.go:225","timestamp":"2022-05-17T11:37:17.990751Z"}
> {"component":"virt-handler","level":"info","msg":"Starting a device plugin for device: example.com/2080_Ti","pos":"device_controller.go:92","timestamp":"2022-05-17T11:37:17.990724Z"}
> {"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A]","pos":"device_controller.go:226","timestamp":"2022-05-17T11:37:17.990793Z"}
> {"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:37:17.993336Z"}
> {"component":"virt-handler","level":"info","msg":"device socket file for device example.com/2080_Ti was removed, kubelet probably restarted.","pos":"pci_device.go:308","timestamp":"2022-05-17T11:37:17.998289Z"}
> {"component":"virt-handler","level":"info","msg":"example.com/2080_Ti device plugin started","pos":"pci_device.go:158","timestamp":"2022-05-17T11:37:19.001290Z"}
> ```
> 
From the logs, it seems that one of device plugin has been disabled:
```
{"component":"virt-handler","level":"info","msg":"disabled device-pluings for: [nvidia.com/TU102_GeForce_RTX_2080_Ti_Rev_A]","pos":"device_controller.go:226","timestamp":"2022-05-17T11:37:17.990793Z"}
```

> I'm using the KubeVirt CRD, the configmaps present are
> 
> ```
> kube-root-ca.crt
> kubevirt-ca
> kubevirt-install-strategy-jm7m2
> ```
> 
> which I do not think overlap with the CRD.
> 
No, this seems fine

> so @alicefr, you don't think the problem has any relevance to the fact that my GPUs are in the same IOMMU group and are identical? I have many other nodes in the cluster (that do not have this condition) and they (and their GPUs) are behaving very nicely.

Probably, can you change the iommu group? I wanted to check any possible conflicts with the configmap as it already happened in the past. 
FYI @vladikr

**Comment by alizshow** [comment]
> Probably, can you change the iommu group?
I tried doing this, the RTX2080Ti is a big card, and there is only one arrangement of the cards on the board for me (although I suspect the board has one PCIe bridge only).

The main problem is that, the device plugin approach cannot work in this situation, since if it registers the 2 cards as allocatable resources on the node, it would mean that they could be used independently, which is not the case since both cards are on the same iommu group, and thus can only be passed to one machine.

What I was hoping to find was that is there a more manual way for passing the cards to the machine, like we would with qemu-kvm?

**Comment by alicefr** [comment]
oh, yes I see. Let me rephrase if I understand you correctly, you'd like to have a device plugin per IOMMU group, right? If that's the case, then I think right now the KubeVirt device plugin only identifies the devices per vendor/product id.

**Comment by alizshow** [comment]
I had tried it with a single resource name, it resulted in only one of GPUs becoming allocatable (and show up on the node capacity).

In the end I had to use the [ACS Override Kernel](https://queuecumber.gitlab.io/linux-acs-override/) patch, to force different IOMMU groups (although theoretically not needed).

**Comment by dhruvik7** [comment]
Is there still no way to pass through multiple PCI devices that are the identical and in the same IOMMU group?

**Comment by pingbird** [comment]
This doesn't appear to be fixed

**Comment by victortoso** [comment]
> What I was hoping to find was that is there a more manual way for passing the cards to the machine, like we would with qemu-kvm?

In your case, how would you do it with qemu-kvm?

**Comment by victortoso** [comment]
/reopen

I'm not sure if current device-plugin would be able to be changed and support this, or if another device-plugin would be needed. If you know other device-plugins that does what you want, it might be possible to use that and minor changes would be needed in KubeVirt's config to accommodate it.

**Comment by dhruvik7** [comment]
The current device plugin (https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/device-manager/pci_device.go) stores a 1:1 map of iommu group to device address, could that just be expanded to be a 1:many map, so it could register all devices in the group?

**Comment by henviso** [comment]
Hey @victortoso.

I work with Dhruvik and [this PR](https://github.com/kubevirt/kubevirt/pull/11606) implements the changes that he is suggesting here we'll be experimenting locally to check whether it solves the part of the issue that is affecting us.

**Comment by henviso** [comment]
> The current device plugin (https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/device-manager/pci_device.go) stores a 1:1 map of iommu group to device address, could that just be expanded to be a 1:many map, so it could register all devices in the group?

Hey @victortoso.

I work with dhruvik7 and [this PR](https://github.com/kubevirt/kubevirt/pull/11606) implements the changes that he is suggesting here we'll be experimenting locally to check whether it solves the part of the issue that is affecting us. 

It is based on [v1.2.0-beta.0](https://github.com/kubevirt/kubevirt/releases/tag/v1.2.0-beta.0) because that's the version we are currently using in our experimental cluster.

**Comment by victortoso** [comment]
Thanks @henviso, I promise to take a look this afternoon. No problem in being based in v1.2.0, shouldn't be hard to rebase on `main` branch.

**Comment by henviso** [comment]
Hello @victortoso.

I closed the draft I mentioned here, because we tried and it didn't work properly.
We just finished testing a variation of this draft change that worked really well for our case, we were able to use 4 different devices under the same iommuGroup with this change.

I'll write another Draft with proper evidence in the morning.

**Comment by henviso** [comment]
Hey @victortoso here is the other draft for the solution that effectively worked in our case: https://github.com/kubevirt/kubevirt/pull/11620.

In short terms we were to fix the existing pci_device plugin with small strategic changes, they are not the cleanest, but might make sense to avoid changing a lot of the stuff surrounding it.

Please, let me know what you think.
